README.md
# NAME

Vector::QRCode::EPS - A generator class for vector data of QRCode

# SYNOPSIS

    use Vector::QRCode::EPS;
    
    my $ps = Vector::QRCode::EPS->generate(
        text        => 'Hello, world!',
        colour      => [255, 0, 0], 
        bgcolour    => [150, 150, 150],
        transparent => 0,
        size        => 6,
        unit        => 'cm',
        qrcode_options => {
            version => 5,
            level   => 'H',
        },
    );
    $ps->output('qrcode.ps');

# DESCRIPTION

Vector::QRCode::EPS is a generator that returns a QRCode data as [PostScript::Simple](https://metacpan.org/pod/PostScript::Simple) object.

# REQUIREMENT

You have to install [libqrencode](https://github.com/fukuchi/libqrencode) into your host before installing this module.

# METHODS

## generate

Returns a [PostScript::Simple](https://metacpan.org/pod/PostScript::Simple) object that contains a vector data of QRCode.

    $ps_obj = Vector::QRCode::EPS->generate(%options);

Options are followings.

- text

    Required. Text that will be implemented into QRCode.

- size

    Optional. Multiple of unit. Default is 10.

- unit

    Optional. Unit from 'mm', 'cm', 'in', 'pt', and 'bp'. Default is 'cm'.

    Please see more datail for [CONSTRUCTOR Paragraph of the PostScript::Simple documentation](http://search.cpan.org/perldoc?PostScript::Simple#CONSTRUCTOR).

- colour

    Optional. RGB colour specification in arrayref. Default is \[10, 10, 10\].

- bgcolour

    Optional. RGP colour specification for background color in arrayref. Default is \[255, 255, 255\].

- transparent

    Optional. Transparent background when true value is specified. Default is undef.

- qrcode\_options

    Optional. Options as [Text::QRCode](https://metacpan.org/pod/Text::QRCode). Default is undef.

# LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

ytnobody <ytnobody@gmail.com>

# SEE ALSO

[libqrencode](https://github.com/fukuchi/libqrencode)

[PostScript::Simple](https://metacpan.org/pod/PostScript::Simple)

[Text::QRCode](https://metacpan.org/pod/Text::QRCode)


Changes
Revision history for Perl extension Vector-QRCode-EPS

{{$NEXT}}

0.08 2017-02-19T03:11:32Z
    - Fixed a test-utility to support a setup section in generated postscript.
      (Thanks to Josef Kutej and Wooter)

0.07 2014-08-05T07:13:00Z
    - more loose testing about cell accuracy

0.06 2014-08-04T15:34:56Z
    - refactor cell tests

0.05 2014-06-25T04:58:18Z
    - change dependency from PostScript::Easy to PostScript::Simple v0.08.
      Because, PostScript::Simple was fixed a bug about providing some Test::* modules in version 0.08.
    - fixed test to use Test::Differences.

0.04 2014-06-04T09:02:44Z
    - change dependency from PostScript::Simple to PostScript::Easy.
      Because, PostScript::Simple provides Test::Simple 0.18 forcely under Carton.

0.03 2014-05-07T09:47:27Z
    - remove unwanted work in test
    - add "qrcode_options" attribute

0.02 2014-03-14T08:36:51Z
    - add background color and transparent feature

0.01 2014-02-28T07:02:15Z

    - original version



t/EPS/Util.pm
package
    t::EPS::Util;
use strict;
use warnings;

sub new {
    my ($class, $string) = @_;
    bless {string => $string}, $class;
}

sub header {
    my ($class, $key) = @_;
    my ($rtn) = $class->{string} =~ /(^\%\!PS\-Adobe\-.*\n\%\%EndComments\n)/ms;
    return defined $key ? [$rtn =~ /^\%\%$key: (.+)$/m]->[0] : $rtn;
}

sub prolog {
    my ($class) = @_;
    my ($rtn) = $class->{string} =~ /(^\%\%BeginProlog.*\%\%EndProlog\n)/ms;
    return $rtn;
}

sub body {
    my ($class) = @_;
    my ($data) = $class->{string} =~ /\%\%EndProlog\n(.*)\%\%EOF/ms;
    my @lines = split /\n/, $data;
    return @lines;
}

1;


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Vector-QRCode-EPS',
    module_name     => 'Vector::QRCode::EPS',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Vector/QRCode/EPS.pm
package Vector::QRCode::EPS;
use 5.008005;
use strict;
use warnings;
use PostScript::Simple;
use Text::QRCode;

our $VERSION = "0.07";

sub generate {
    my ($class, %opts) = @_;
    my $text = $opts{text};
    my $size = $opts{size} || 10;
    my $unit = $opts{unit} || 'cm';
    my $colour = $opts{colour} || [10, 10, 10];
    my $bgcolour = $opts{bgcolour} || [255,255,255];
    my $transparent = $opts{transparent};

    my $qrcode_options = $opts{qrcode_options} || {};

    my $qrdata = Text::QRCode->new(%$qrcode_options)->plot($text);
    $qrdata = [reverse @$qrdata]; ### avoid to upside down

    my $xsize = scalar( @{$qrdata->[0]} );
    my $ysize = scalar( @{$qrdata} );

    my $ps = PostScript::Simple->new(
        colour => 1, 
        eps => 1, 
        units => $unit, 
        xsize => $size, 
        ysize => $size
    );

    unless ($transparent) {
        $ps->setcolour(@$bgcolour);
        $ps->box({filled => 1}, 0, 0, $size, $size);
    }

    $ps->setcolour(@$colour);
    my ($x, $y);
    for $y (0 .. $#{$qrdata}) {
        for $x (0 .. $#{$qrdata->[$y]} ) {
            if ($qrdata->[$y][$x] eq '*') {
                $ps->box(
                    {filled => 1}, 
                    $size/$xsize*$x, $size/$ysize*$y, 
                    $size/$xsize*($x+1), $size/$ysize*($y+1)
                );
            }
        }
    }

    $ps;
}

1;
__END__

=encoding utf-8

=head1 NAME

Vector::QRCode::EPS - A generator class for vector data of QRCode

=head1 SYNOPSIS

    use Vector::QRCode::EPS;
    
    my $ps = Vector::QRCode::EPS->generate(
        text        => 'Hello, world!',
        colour      => [255, 0, 0], 
        bgcolour    => [150, 150, 150],
        transparent => 0,
        size        => 6,
        unit        => 'cm',
        qrcode_options => {
            version => 5,
            level   => 'H',
        },
    );
    $ps->output('qrcode.ps');


=head1 DESCRIPTION

Vector::QRCode::EPS is a generator that returns a QRCode data as L<PostScript::Simple> object.

=head1 REQUIREMENT

You have to install L<libqrencode|https://github.com/fukuchi/libqrencode> into your host before installing this module.

=head1 METHODS

=head2 generate

Returns a L<PostScript::Simple> object that contains a vector data of QRCode.

    $ps_obj = Vector::QRCode::EPS->generate(%options);

Options are followings.

=over 4

=item text

Required. Text that will be implemented into QRCode.

=item size

Optional. Multiple of unit. Default is 10.

=item unit

Optional. Unit from 'mm', 'cm', 'in', 'pt', and 'bp'. Default is 'cm'.

Please see more datail for L<CONSTRUCTOR Paragraph of the PostScript::Simple documentation|http://search.cpan.org/perldoc?PostScript::Simple#CONSTRUCTOR>.

=item colour

Optional. RGB colour specification in arrayref. Default is [10, 10, 10].

=item bgcolour

Optional. RGP colour specification for background color in arrayref. Default is [255, 255, 255].

=item transparent

Optional. Transparent background when true value is specified. Default is undef.

=item qrcode_options

Optional. Options as L<Text::QRCode>. Default is undef.

=back

=head1 LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ytnobody E<lt>ytnobody@gmail.comE<gt>

=head1 SEE ALSO

L<libqrencode|https://github.com/fukuchi/libqrencode>

L<PostScript::Simple>

L<Text::QRCode>

=cut

